File: datestructure/javascript/interview-2-hashtable/firstRecurringCharacter.js
//Google Question
//Given an array = [2,5,1,2,3,5,1,2,4];
//It should return 2

//Given an array = [2,1,1,2,3,5,1,2,5]
//It should return 1

//Given an array = [2,3,4,5]
//It should return undefined

function firstRecurringCharacter(input) {
    for (i = 0; i < input.length; i++) {
        for ( j = 0; j < i; j++) {
            const temp = input[i]
            if ( temp === input[j]) {
                return temp
            }
        }
    }
    return undefined
}

function firstRecurringCharacter2(input) {
    let map = new Map()
    for (i = 0; i < input.length; i++) {
        if(input[i] === map.get(input[i])) {
            return input[i]
        }
        map.set(input[i],input[i])
    }
    console.log(map.size)
    return undefined
}

function firstRecurringCharacter3(input) {
    let map = {}
    for (i = 0; i < input.length; i++) {
        if ( map[input[i]] === input[i]) {
            return input[i]
        }
        map[input[i]] = input[i]
    }

    return undefined

}

var map = {}
map.key = 1
map.key = 2
map[1] = 1
console.log(map[1])

const str1 = [2,5,1,2,3,5,1,2,4];
const str2 = [2,1,1,2,3,5,1,2,5];
const str3 = [2,3,4,5];

console.log(firstRecurringCharacter3(str3));

File: datestructure/java/interview-2-hashTab/FirstRecurringCharacter.java
import java.util.Hashtable;

//Google Question
//Given an array = [2,5,1,2,3,5,1,2,4];
//It should return 2

//Given an array = [2,1,1,2,3,5,1,2,5]
//It should return 1

//Given an array = [2,3,4,5]
//It should return undefined

class FirstRecurringCharacter {

    public static void main(String[] args) {
        int[] arr = {2,5,1,2,3,5,1,2,4};
        int[] arr1 = {2,1,1,2,3,5,1,2,4};
        int[] arr2 = {2,3,4,5};

        System.out.println(firstRecurringCharacter(arr2));
    }

    public static int firstRecurringCharacter(int[] arr) {
        Hashtable<Integer, Integer> hashTab = new Hashtable<>();
        int number = -1;
        for (int i = 0; i < arr.length; i++) {
            if(hashTab.containsKey(arr[i])){
                return arr[i];
            }
            hashTab.put(arr[i], arr[i]);
        }
        return number;

    }
}